File: npm/typescript_oxlint/ts/rules/prefer_regexp_exec.ts
import { calleePropertyName, regexFlags } from "./ast";
import { createNativeRule } from "./rule_creator";

export const preferRegexpExecRule = createNativeRule(
  "prefer-regexp-exec",
  {
    docs: {
      description: "Prefer RegExp#exec over String#match for single matches.",
    },
    messages: {
      unexpected: "Use a RegExp exec() call instead of String match().",
    },
  },
  (context) => ({
    CallExpression(node: any) {
      if (calleePropertyName(node) !== "match") {
        return;
      }
      const [firstArgument] = node.arguments;
      if (!firstArgument) {
        return;
      }
      const flags = regexFlags(firstArgument);
      if (flags !== undefined && !flags.includes("g")) {
        context.report({ node, messageId: "unexpected" });
      }
    },
  }),
);

File: src/core/corsa_core/src/lint/rules/prefer_regexp_exec.rs
use super::super::{LintNode, RuleContext, RuleMessage, RustLintRule};
use crate::lint::helpers::{callee_property_name, first_child_list_item, regex_flags};

#[derive(Clone, Copy, Debug, Default)]
pub struct PreferRegexpExecRule;

const MESSAGES: &[RuleMessage] = &[RuleMessage {
    id: "unexpected",
    description: "Use a RegExp exec() call instead of String match().",
}];
const LISTENERS: &[&str] = &["CallExpression"];

impl RustLintRule for PreferRegexpExecRule {
    fn name(&self) -> &'static str {
        "prefer-regexp-exec"
    }

    fn docs_description(&self) -> &'static str {
        "Prefer RegExp#exec over String#match for single matches."
    }

    fn messages(&self) -> &'static [RuleMessage] {
        MESSAGES
    }

    fn listeners(&self) -> &'static [&'static str] {
        LISTENERS
    }

    fn requires_type_texts(&self) -> bool {
        false
    }

    fn check(&self, ctx: &mut RuleContext<'_>, node: &LintNode) {
        if node.kind != "CallExpression"
            || callee_property_name(Some(node)).as_deref() != Some("match")
        {
            return;
        }
        let Some(first_argument) = first_child_list_item(node, "arguments") else {
            return;
        };
        let Some(flags) = regex_flags(first_argument) else {
            return;
        };
        if !flags.contains('g') {
            ctx.report("unexpected", node.range);
        }
    }
}
